packages/client-db/src/Sync/index.ts

```diff
@@ -0,0 +1,154 @@
+
+import path from 'path';
+import { Worker } from 'worker_threads';
+import logger from '@polkadot/util/logger';
+import promisify from '@polkadot/util/promisify';
+// import u8aToHex from '@polkadot/util/u8a/toHex';
+
+import commands from './commands';
+
+const l = logger('db/main');
+
+const U32_MAX = Math.pow(2, 32) - 1;
+const returnable = ['get', 'root'];
+
+export default class SyncDb implements TrieDb {
+  private child: WorkerThreads.Worker;
+
+  constructor () {
+    // NOTE With --experimental-modules (ESNext targets) this can become
+    // this.child = new Worker(new URL('./worker.js', import.meta.url).pathname);
+    this.child = new Worker(path.join(__dirname, './worker.js'));
+  }
+
+  // Sends a message to the worker, optionally (in the case of get/root) returning the
+  // actual result.
+  private _sendMessage (type: MessageType, key?: Uint8Array, value?: Uint8Array): Uint8Array | null {
+    const state = new Int32Array(new SharedArrayBuffer(8));
+    let view: DataView | null = null;
+    let buffer: Uint8Array | null = null;
+
+    const start = () => {
+      state[0] = commands.START;
+      this.child.postMessage({
+        buffer,
+        key,
+        state,
+        value,
+        type
+      } as Message);
+
+      Atomics.wait(state, 0, commands.START);
+    };
+
+    // Ok, this is not something that returns a value, just send the message and
+    // return when we the call has been done
+    if (!returnable.includes(type)) {
+      start();
+      return null;
+    }
+
+    // The shared data buffer that will be used by the worker to send info back
+    // (wrapped in Uint8Array for binary data or in DataView for Int32 data)
+    const shared = new SharedArrayBuffer(4096);
+
+    buffer = new Uint8Array(shared);
+    view = new DataView(shared);
+
+    let result: Uint8Array | null = null;
+    let size = 0;
+    let offset = 0;
+
+    start();
+
+    // Here we loop through the states and either read data to fill the buffer
+    // or return when it is time to do so
+    while (true) {
+      switch (state[0]) {
+        // we have reached the end, return what we have
+        case commands.END:
+          return result;
+
+        // Error, so just return a null for the caller to handle
+        case commands.ERROR:
+          return null;
+
+        // Ahah, we need to read the size (first result) to detemine how
+        // big of a result buffer we need.
+        case commands.SIZE:
+          size = view.getUint32(0);
+
+          if (size === U32_MAX) {
+            return null;
+          }
+
+          result = new Uint8Array(size);
+
+          this._notifyFill(state);
+          break;
+
+        // Get the available data from the buffer and write it into our result
+        // array.
+        case commands.READ:
+          const available = Math.min(buffer.length, size - offset);
+
+          (result as Uint8Array).set(buffer.subarray(0, available), offset);
```

Comment:
If for whatever reason you get `READ` before `SIZE` this will obviously error. Is there any scenario in which the first expected command isn't `SIZE`? I'd rather split the switch case into two:

1) switch on command expecting  `ERROR | SIZE`
2) switch in a loop expecting `ERROR | END | READ`, `result` here is guaranteed to be non-`null`.

Thus your state machine won't have invalid transitions (`result` is uninitialized but `READ` was read, or `result` is initialized but `SIZE` was read again).